crates/wasm-smith/src/core.rs

```diff
@@ -509,6 +561,201 @@ impl Module {
+                            func_type,
+                        };
+                        let entity = EntityType::Tag(tag_type.clone());
+                        if type_size_budget < entity.size() {
+                            continue;
+                        }
+                        self.tags.push(tag_type);
+                        entity
+                    } else {
+                        continue;
+                    }
+                }
+
+                wasmparser::TypeRef::Table(table_ty) => {
+                    let table_ty = TableType {
+                        element_type: convert_type(table_ty.element_type),
+                        minimum: table_ty.initial,
+                        maximum: table_ty.maximum,
+                    };
+                    let entity = EntityType::Table(table_ty);
+                    let type_size = entity.size();
+                    if type_size_budget < type_size || !self.can_add_local_or_import_table() {
+                        continue;
+                    }
+                    self.type_size += type_size;
+                    self.tables.push(table_ty);
+                    entity
+                }
+
+                wasmparser::TypeRef::Memory(memory_ty) => {
+                    let memory_ty = MemoryType {
+                        minimum: memory_ty.initial,
+                        maximum: memory_ty.maximum,
+                        memory64: memory_ty.memory64,
+                    };
+                    let entity = EntityType::Memory(memory_ty.clone());
+                    let type_size = entity.size();
+                    if type_size_budget < type_size || !self.can_add_local_or_import_memory() {
+                        continue;
+                    }
+                    self.type_size += type_size;
+                    self.memories.push(memory_ty);
+                    entity
+                }
+
+                wasmparser::TypeRef::Global(global_ty) => {
+                    let global_ty = GlobalType {
+                        val_type: convert_type(global_ty.content_type),
+                        mutable: global_ty.mutable,
+                    };
+                    let entity = EntityType::Global(global_ty);
+                    let type_size = entity.size();
+                    if type_size_budget < type_size || !self.can_add_local_or_import_global() {
+                        continue;
+                    }
+                    self.type_size += type_size;
+                    self.globals.push(global_ty);
+                    entity
+                }
+            };
+            new_imports.push(Import {
+                module: import.module.to_string(),
+                field: import.name.to_string(),
+                entity_type,
+            });
+            self.num_imports += 1;
+        }
+
+        // Finally, add the sections we just generated.
+        if !new_imports.is_empty() || u.arbitrary().unwrap_or(false) {
```

Comment:
Is `u.arbitrary()` call effectively only making it so we add all available imports or none of them? It seems to me like we should be calling `u.arbitrary()` (or `u.ratio(...)`) for each import instead.

Reply:
The `arbitrary` here makes a decision as to whether an empty section should be added or not. This is a pattern that is replicated throughout the this module, so perhaps it calls for some sort of abstraction.